### DISCUSSION TITLE
Node Version Not Updated in Global Config After Upgrade

### DISCUSSION THREAD

**Discussion by itizarsa** [jdx/mise#3842] [Q&A]
**Describe the bug**
When upgrading Node.js using `mise upgrade node@22` or `node@22.10.0`, the new version is installed correctly, but the global configuration is not updated to reflect the new version. This causes mise to emit a "mise WARN  missing: node@22.10.0"

**To Reproduce**
1. `mise use -g node@22.10.0`
2. Run `mise upgrade node@22` or `mise upgrade node@22.10.0`
3. Open a new shell session

**Expected behavior**
- The new Node.js version should be installed
- Global configuration should be updated automatically
- New shell sessions should use the upgraded Node.js version

**`mise doctor` output**

```text
version: 2024.12.20 macos-arm64 (2024-12-25)
activated: no
shims_on_path: yes

build_info:
  Target: aarch64-apple-darwin
  Features: DEFAULT, NATIVE_TLS
  Built: Wed, 25 Dec 2024 20:45:03 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26) (Homebrew)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin23.0)

dirs:
  cache: ~/Library/Caches/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/.config/mise/config.toml

ignored_config_files: (none)

backends:
  aqua
  asdf
  cargo
  core
  dotnet
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:
  jbang  https://github.com/jbangdev/jbang-asdf.git#96c41a7

toolset:
  core:bun@1.1.31
  core:deno@2.0.2
  core:go@1.23.2
  core:java@liberica-21.0.5+11
  core:node@22.10.0     (missing)
  core:python@3.13.1

path:
  /opt/homebrew/bin
  /opt/homebrew/sbin
  /usr/local/bin
  /System/Cryptexes/App/usr/bin
  /usr/bin
  /bin
  /usr/sbin
  /sbin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin
  /var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
  ~/.local/share/mise/installs/bun/1.1.31/bin
  ~/.local/share/mise/installs/java/liberica-21.0.5+11/bin
  ~/.local/share/mise/installs/deno/2.0.2/bin
  ~/.local/share/mise/installs/deno/2.0.2/.deno/bin
  ~/.local/share/mise/installs/go/1.23.2/bin
  ~/.local/share/mise/installs/python/3.13.1/bin
  ~/.orbstack/bin
  ~/.orbstack/bin

env_vars:
  (none)

settings:

No problems found
```

**Comment by syhol** [comment]
I did notice your mise doctor says `  core:node@22.10.0     (missing)`, this tells me:
- the new version does not appear to be installed, the "missing" meaning it's not installed
- It did upgrade your config (not sure if its global or local) to your requested version of `22.10.0`

I think I would need a bit more to help you any further:

1. What was the full output of the command? can you run it with verbose flags `-vv`?
2. Did you have any mise config files (other than the user level ~/.config/mise/config.toml) that could have been upgraded instead?

**Comment by Marukome0743** [comment]
The important distinction here is that an exact version such as `node = "22.10.0"` is treated as a pin. A normal `mise upgrade` preserves the configured request rather than silently rewriting an exact pin.

If the goal is to update the global config to the latest available Node 22 release while continuing to store an exact version, the current command is:

```sh
mise use -g --pin node@22
```

`node@22` limits resolution to the Node 22 series, `--pin` writes the resolved exact version, and `-g` targets the global config.

Alternatively, if exact version strings are not required in the config, use:

```sh
mise use -g node@22
mise upgrade node
```

The global config then remains `node = "22"`, while `mise upgrade node` installs the newest matching Node 22 release. A lockfile can be enabled when reproducibility is needed without storing the exact patch version directly in `config.toml`.

`mise upgrade --bump node` has different semantics: it permits changing the configured version range and may move to a newer major release.

I also verified on the current main branch that config bump operations update both `mise.toml` and `.tool-versions` correctly while preserving unrelated settings and comments.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
